lib/DxilValidation/DxilValidation.cpp

```diff
@@ -970,6 +970,226 @@ static void ValidateImmOperandForMathDxilOp(CallInst *CI, DXIL::OpCode Opcode,
+                                          bool Transposed) {
+  switch (Layout) {
+  case DXIL::DXILMatrixLayout::RowMajor:
+  case DXIL::DXILMatrixLayout::ColumnMajor:
+    return !Transposed;
+
+  default:
+    return true;
+  }
+}
+
+static bool CheckUnsignedFlag(Type *VecTy, bool IsUnsigned) {
+  Type *ElemTy = VecTy->getScalarType();
+  if (ElemTy->isFloatingPointTy())
+    return !IsUnsigned;
+
+  return true;
+}
+
+static Value *getMatVecOpIsOutputUnsigned(CallInst *CI, DXIL::OpCode OpCode) {
+  switch (OpCode) {
+  case DXIL::OpCode::MatVecMul:
+    return CI->getOperand(DXIL::OperandIndex::kMatVecMulIsOutputUnsignedIdx);
+  case DXIL::OpCode::MatVecMulAdd:
+    return CI->getOperand(DXIL::OperandIndex::kMatVecMulAddIsOutputUnsignedIdx);
+
+  default:
+    DXASSERT_NOMSG(false);
+    return nullptr;
+  }
+}
+
+static void ValidateImmOperandsForMatVecOps(CallInst *CI, DXIL::OpCode OpCode,
+                                            ValidationContext &ValCtx) {
+
+  // Check Common operands
+  llvm::Value *IsInputUnsigned =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulIsInputUnsignedIdx);
+  llvm::Value *InputInterpretation =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulInputInterpretationIdx);
+  llvm::Value *MatrixInterpretation =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixInterpretationIdx);
+  llvm::Value *MatrixM =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixMIdx);
+  llvm::Value *MatrixK =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixKIdx);
+  llvm::Value *MatrixLayout =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixLayoutIdx);
+  llvm::Value *MatrixTranspose =
+      CI->getOperand(DXIL::OperandIndex::kMatVecMulMatrixTransposeIdx);
+  llvm::Value *IsOutputUnsigned = getMatVecOpIsOutputUnsigned(CI, OpCode);
+
+  ConstantInt *IsInputUnsignedConst =
+      dyn_cast<llvm::ConstantInt>(IsInputUnsigned);
+  if (!IsInputUnsignedConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+    return;
+  }
+
+  ConstantInt *IsOutputUnsignedConst =
+      dyn_cast<llvm::ConstantInt>(IsOutputUnsigned);
+  if (!IsOutputUnsignedConst) {
+    ValCtx.EmitInstrError(CI,
+                          ValidationRule::InstrMatVecOpIsUnsignedFlagsAreConst);
+    return;
+  }
+
+  ConstantInt *II = dyn_cast<ConstantInt>(InputInterpretation);
+  ConstantInt *MI = dyn_cast<ConstantInt>(MatrixInterpretation);
+  if (!II || !MI) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInterpretationParamAreConst);
+    return;
+  }
+
+  // Check if InputInterpretation and MatrixInterpretation are valid
+  auto IIValue = II->getLimitedValue();
+  if (!CheckFromRegisterInterpretations(IIValue)) {
+    ValCtx.EmitInstrError(
+        CI, ValidationRule::InstrLinalgInvalidRegisterInterpValue);
+    return;
+  }
+
+  auto MIValue = MI->getLimitedValue();
```

Comment:
nit: https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable
```suggestion
  uint64_t MIValue = MI->getLimitedValue();
```